openwisp_controller/config/api/views.py

```diff
@@ -0,0 +1,110 @@
+from rest_framework import pagination
+from rest_framework.authentication import SessionAuthentication
+from rest_framework.generics import (
+    ListCreateAPIView,
+    RetrieveAPIView,
+    RetrieveUpdateDestroyAPIView,
+)
+from rest_framework.permissions import DjangoModelPermissions, IsAuthenticated
+from swapper import load_model
+
+from openwisp_users.api.authentication import BearerAuthentication
+from openwisp_users.api.mixins import FilterByOrganizationManaged
+
+from ..admin import BaseConfigAdmin
+from .serializers import (
+    DeviceDetailSerializer,
+    DeviceListSerializer,
+    TemplateSerializer,
+    VpnSerializer,
+)
+
+Template = load_model('config', 'Template')
+Vpn = load_model('config', 'Vpn')
+Device = load_model('config', 'Device')
+Config = load_model('config', 'Config')
+
+
+class ListViewPagination(pagination.PageNumberPagination):
+    page_size = 10
+    page_size_query_param = 'page_size'
+    max_page_size = 100
+
+
+class ProtectedAPIMixin(FilterByOrganizationManaged):
+    authentication_classes = [BearerAuthentication, SessionAuthentication]
+    permission_classes = [
+        IsAuthenticated,
+        DjangoModelPermissions,
+    ]
+
+
+class TemplateListCreateView(ProtectedAPIMixin, ListCreateAPIView):
+    serializer_class = TemplateSerializer
+    queryset = Template.objects.order_by('-created')
+    pagination_class = ListViewPagination
+
+
+class TemplateDetailView(ProtectedAPIMixin, RetrieveUpdateDestroyAPIView):
+    serializer_class = TemplateSerializer
+    queryset = Template.objects.all()
+
+
+class DownloadTemplateconfiguration(ProtectedAPIMixin, RetrieveAPIView):
+    serializer_class = TemplateSerializer
+    queryset = Template.objects.none()
+    model = Template
+
+    def retrieve(self, request, *args, **kwargs):
+        return BaseConfigAdmin.download_view(self, request, pk=kwargs['pk'])
+
+
+class VpnListCreateView(ProtectedAPIMixin, ListCreateAPIView):
+    serializer_class = VpnSerializer
+    queryset = Vpn.objects.order_by('-created')
+    pagination_class = ListViewPagination
+
+
+class VpnDetailView(ProtectedAPIMixin, RetrieveUpdateDestroyAPIView):
+    serializer_class = VpnSerializer
+    queryset = Vpn.objects.all()
+
+
+class DownloadVpnView(ProtectedAPIMixin, RetrieveAPIView):
+    serializer_class = VpnSerializer
+    queryset = Vpn.objects.none()
+    model = Vpn
+
+    def retrieve(self, request, *args, **kwargs):
+        return BaseConfigAdmin.download_view(self, request, pk=kwargs['pk'])
+
+
+class DeviceListCreateView(ProtectedAPIMixin, ListCreateAPIView):
+    serializer_class = DeviceListSerializer
+    queryset = Device.objects.order_by('-created')
```

Comment:
we may need to add `.select_related('config')` here too

Reply:
@ManishShah120 did you see one of my previous reviews I mentioned the following?

> It may be important to show also config.status and config.backend in the list, since these two pieces of information are quite important. Can you add a limited config serializer in the list serializer too @ManishShah120? (Keep in mind to use select_related('config') to keep the number of generated queries under control).

Did you see this?

Yes, this has been implemented, I was able to achieve this without introducing a new serializer, I introduced two new method fields and got its value by checking if an device has config with `._has_config()` and returned their status, and backend value.
```
status = serializers.SerializerMethodField()
backend = serializers.SerializerMethodField()
```